Radix client

balda-rdx opened the pull request:
This PR adds a Client for the radix network.

This client makes use of the radix gateway api: https://radix-babylon-gateway-api.redoc.ly/
It also uses the radix engine toolkit: https://github.com/radixdlt/typescript-radix-engine-toolkit/tree/main and the gateway sdk: https://www.npmjs.com/package/@radixdlt/babylon-gateway-api-sdk

0xp3gasus commented on packages/xchain-radix/package.json:
```diff
@@ -0,0 +1,55 @@
+  },
+  "scripts": {
+    "clean": "rm -rf .turbo && rm -rf lib",
+    "build": "yarn clean && rollup -c",
```
Add the "build:release" script the rest of the clients have

balda-rdx replied:
done

0xp3gasus commented:
> This PR adds a Client for the radix network.
> 
> This client makes use of the radix gateway api: https://radix-babylon-gateway-api.redoc.ly/ It also uses the radix engine toolkit: https://github.com/radixdlt/typescript-radix-engine-toolkit/tree/main and the gateway sdk: https://www.npmjs.com/package/@radixdlt/babylon-gateway-api-sdk

First of all, thanks for opening the pull request!! I left you a few comments due the latest changes we have made in the repo. On the other side. Is there any way to get testnet faucet to test transactions?

balda-rdx commented:
> > This PR adds a Client for the radix network.
> > This client makes use of the radix gateway api: https://radix-babylon-gateway-api.redoc.ly/ It also uses the radix engine toolkit: https://github.com/radixdlt/typescript-radix-engine-toolkit/tree/main and the gateway sdk: https://www.npmjs.com/package/@radixdlt/babylon-gateway-api-sdk
> 
> First of all, thanks for opening the pull request!! I left you a few comments due the latest changes we have made in the repo. On the other side. Is there any way to get testnet faucet to test transactions?

@0xp3gasus thanks for the feedback. I'll be addressing all the comments and yes, there's a way to get testnet funds from the faucet. will add it

balda-rdx commented on packages/xchain-radix/README.md:
```diff
@@ -0,0 +1,207 @@
+# `@xchainjs/xchain-client`
```
done

balda-rdx commented on packages/xchain-radix/package.json:
```diff
@@ -0,0 +1,55 @@
+    "bech32": "^2.0.0",
+    "bip32": "^4.0.0",
+    "ed25519-hd-key": "^1.3.0",
+    "rollup-plugin-peer-deps-external": "^2.2.4"
```
I think it was needed with the previous structure of the repo, but might not be needed anymore. removing it

0xp3gasus commented on packages/xchain-radix/src/client.ts:
```diff
@@ -0,0 +1,852 @@
+export default class Client extends BaseXChainClient {
+  private radixSpecificClient: RadixSpecificClient
+  private curve: Curve
+  constructor(
```
To be consistent with the rest of XChainJS clients, use the following type for the constructor `XChainClientParams & { curve?: Curve }`. To simplify the creation of the client, I would add 'Ed25519' as default value to the curve parameter.

balda-rdx replied:
done

0xp3gasus commented on packages/xchain-radix/src/client.ts:
```diff
@@ -0,0 +1,852 @@
+   * @returns {Address} A promise resolving to the current address.
+   * A phrase is needed to create a wallet and to derive an address from it.
+   */
+  async getAddressAsync(): Promise<string> {
```
Add walletIndex parameter to `getAddressAsync`

0xp3gasus commented on packages/xchain-radix/src/client.ts:
```diff
@@ -0,0 +1,852 @@
+   *
+   * @returns {Fee} An estimated fee
+   */
+  async getFees(): Promise<Fees> {
```
I could not test this function in Mainnet network. This is the error I have

```ts
    Preview for fees was not successful

      265 |
      266 |     if (previewReceipt.receipt.status !== 'Succeeded') {
    > 267 |       throw new Error('Preview for fees was not successful')
          |             ^
      268 |     }
```

0xp3gasus commented on packages/xchain-radix/src/client.ts:
```diff
@@ -0,0 +1,852 @@
+        new PublicKey.Ed25519(
+          Convert.HexString.toUint8Array(feesEstimationPublicKeys[this.getRadixNetwork()].publicKey),
+        ),
+        'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa ...
```
Why is this needed? Does the message affect the fee estimation?

balda-rdx replied:
yes the message length does affect the fee estimation

0xp3gasus commented on packages/xchain-radix/src/client.ts:
```diff
@@ -0,0 +1,852 @@
+  ): Balance[] {
+    const balances: Balance[] = []
+    // Iterate through nonFungibleResources
+    const decimalDiff = XRD_DECIMAL - 8
```
Why is this decimalDiff needed? AssetAmount and BaseAmount can handle assets with non 8 decimals

balda-rdx replied:
got it

0xp3gasus commented on packages/xchain-radix/src/client.ts:
```diff
@@ -0,0 +1,852 @@
+
+        // We need to do this because item.amount can be either string or number
+        // depending on the type of resource
+        let xrdAmount: number
```
Following the previous comment. This could be simplied as

```ts
       const amount = typeof item.amount === 'string' ? parseFloat(item.amount) : item.amount

        const balance: Balance = {
          asset: asset,
          amount: assetToBase(assetAmount(amount, XRD_DECIMAL)),
        }
```

balda-rdx replied:
makes sense to me

0xp3gasus commented on packages/xchain-radix/src/client.ts:
```diff
@@ -0,0 +1,852 @@
+
+        const balance: Balance = {
+          asset: asset,
+          amount: baseAmount(xrdAmount, XRD_DECIMAL),
```
Does all resources have the same number as decimals?

balda-rdx replied:
not necessarily. fixed

0xp3gasus replied:
Is there a way to retrieve the decimals? For non 8 decimals, the number of decimals should be set, otherwise amount will not be correct

0xp3gasus commented on packages/xchain-radix/src/client.ts:
```diff
@@ -0,0 +1,852 @@
+    return this.innerGatewayClient.status.getCurrent().then((status) => status.ledger_state.state_version)
+  }
+
+  public async fetchBalances(address: string): Promise<Balance[]> {
```
Really interesting feature the non fungible balances. As the other clients do not support it, and to be consistent with them, I would split the code creating two different functions, one for fungible balances which can be retrieve from the getBalance function and another one for the non fungible part

balda-rdx replied:
@0xp3gasus I can totally split it but what would you like the getBalance function to return? just fungible tokens balances? that would be sort of inaccurate as non fungible tokens do contribute to the account balance

0xp3gasus replied:
EVM chains has also NFTs, but only ERC20 balances are returned. I suggest this change to be consistent with them and create a specific feature for NFTs for those clients which are compatible with them, for example, `getNFTBalances`

balda-rdx replied:
fair enough. I'm splitting this

0xp3gasus commented on packages/xchain-radix/src/client.ts:
```diff
@@ -0,0 +1,852 @@
+          txn.confirmed_at !== undefined
+        ) {
+          const transaction: Tx = await this.convertTransactionFromHex(txn.raw_hex, txn.intent_hash, txn.confirmed_at)
+          txList.txs.push(transaction)
```
`total` has to be updated too

0xp3gasus commented on packages/xchain-radix/src/client.ts:
```diff
@@ -0,0 +1,852 @@
+          from_ledger_state: {
+            state_version: params.offset,
+          },
+          manifest_resources_filter: [params.asset || ''],
```
`manifest_resources_filter: params.asset ? [params.asset] : undefined,`

0xp3gasus commented on packages/xchain-radix/src/client.ts:
```diff
@@ -0,0 +1,852 @@
+   * @param {TxHistoryParams} params The options to get transaction history. (optional)
+   * @returns {TxsPage} The transaction history.
+   */
+  async getTransactions(params: TxHistoryParams): Promise<TxsPage> {
```
Is `offset` parameter mandatory?

balda-rdx replied:
I made it optional now

0xp3gasus commented on packages/xchain-radix/package.json:
```diff
@@ -0,0 +1,54 @@
+    "lint": "eslint \"{src,__tests__}/**/*.ts\" --fix --max-warnings 0",
+    "fund": "ts-node src/utils.ts"
+  },
+  "devDependencies": {
```
Remove devDependencies

0xp3gasus reviewed (changes requested):
I could not test transfers in Testnet network due to an error on the getFees

balda-rdx commented:
> I could not test transfers in Testnet network due to an error on the getFees

there's an e2e test that transfer funds in the testnet network, maybe take a look at that one? https://github.com/xchainjs/xchainjs-lib/pull/1146/files#diff-86a40cb25a0073784030ae41e4080be5316ec62d4291712315fd94916e2d1582R404

0xp3gasus commented on packages/xchain-radix/src/client.ts:
```diff
@@ -0,0 +1,853 @@
+
+        const balance: Balance = {
+          asset: asset,
+          amount: assetToBase(assetAmount(amount)),
```
```assetToBase(assetAmount(amount, XRD_DECIMAL))```

0xp3gasus commented on packages/xchain-radix/src/client.ts:
```diff
@@ -157,16 +157,11 @@ export class RadixSpecificClient {
         const balance: Balance = {
           asset: asset,
-          amount: baseAmount(xrdAmount, XRD_DECIMAL),
+          amount: assetToBase(assetAmount(amount)),
```
Is there an easy way to retrieve the decimals from the assets?

balda-rdx replied:
this should be now fixed

0xp3gasus reviewed (approved):
Great work!! 🚀🚀
